flink-runtime/src/main/java/org/apache/flink/migration/v0/SavepointV0.java

```diff
@@ -0,0 +1,207 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.migration.v0;
+
+import org.apache.flink.migration.v0.runtime.TaskStateV0;
+import org.apache.flink.runtime.checkpoint.TaskState;
+import org.apache.flink.runtime.checkpoint.savepoint.Savepoint;
+import org.apache.flink.util.Preconditions;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Savepoint version 0.
+ *
+ * <p>This format was introduced with Flink 1.1.0.
+ * <pre>
+ *     checkpointId: long
+ *     numTaskStates: int
+ *     |----jobVertexID: long[2]
+ *     |----parallelism: int
+ *     |----numSubtaskStates: int
+ *     |    |----subtaskIndex: int
+ *     |    |----serializedValueLength: int
+ *     |    |----serializedValue: byte[] (null if serializedValueLength is -1)
+ *     |    |----subtaskStateSize: long
+ *     |    |----subtaskStateDuration: long
+ *     |----numKeyGroupStates: int
+ *     |    |----subtaskIndex: int
+ *     |    |----serializedValueLength: int
+ *     |    |----serializedValue: byte[] (null if serializedValueLength is -1)
+ *     |    |----keyGroupStateSize: long
+ *     |    |----keyGroupStateDuration: long
+ * </pre>
+ */
+@Deprecated
+@SuppressWarnings("deprecation")
+public class SavepointV0 implements Savepoint {
+
+	/** The classes that are migrated in SavepointV0 */
+	public static final Map<String, String> MigrationMapping = new HashMap<String, String>() {{
+
+		/* migrated state descriptors */
+		put("org.apache.flink.api.common.state.StateDescriptor",
```

Comment:
I agree with your assessment about future changes. Typically, if we have a good testing for backwards compatibility in place, it will also reveal classes that are forgotten in the migration package through `ClassNotFoundException`s - but we should not only rely on that, I see this more like a second level safety-net.

Then I would suggest that we go for option b).